### ISSUE TITLE
Improve how we write rule tests.

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#5803]
I would like to write rule tests like this:

```kotlin
@Test
fun `a description here`() {
  """
    fun function(🐞x: Int🐞) {
      println("Hello World!")
    }
  """.trimIndent()
    .test(UnusedParameter()) // test will know that there is ONE issue between the 🐞 so it can check the count and the position.
}
```

```kotlin
@Test
fun `a description here`() {
  """
    fun function(➡️x: Int⬅️) {
      println("Hello World!")

    }
  """.trimIndent()
    .withDescriptions("Why are you not using `x`?") // vararg of `String` or `Matcher<String>`
    .test(UnusedParameter(), compile = false) // compile the snippet by default. Don't compile is an opt-in parameter
}
```

We should refine the API a bit more and we should define which emoji or marker we should use but I think that something like this would help us A LOT to maintain our tests.

Reasons:
- Right now test start and end of an issue is really difficult. This makes it really easy
- Sometimes, to make some code compile we need to write a lot of boilerplate. In those cases, as a reader, is difficult to spot where is the issue that we say that exists in the code.
- Unify a bit how to write rule tests.

This API should not deprecate the old one. Or, at least, not right away.

What do you think? If you like the idea I can create a PR and implement a first version.

@t-kameyama, @VitalyVPinchuk and @atulgpt (I'm tagging some contributors that usually write/read a lot of tests to get some feedback)

**Comment by TWiStErRob** [comment]
+1 I was really missing this when I wrote my own Detekt rules.

I would add the requirement that at least 99% of the tests should be able to compile if their triple-quoted text is just copied out of the block into real code. Meaning I'm strongly against using emojis.

---

Android Lint did this really well. I would say we should adopt the same API. Here's an example (marked `THIS`):
```kotlin
lint()
    .files(
        // These two are Strings that define the whole file,
        // because third party dependencies need to have at least signatures for type resolution.
        ButterKnife.BindView,
        ButterKnife.OnClick,
        java(
            """
            import androidx.annotation.Nullable;
            import butterknife.BindView;
            import butterknife.OnClick;
            import my.R;
            
            public class MyView {
                @BindView(/*ButterKnife bindings for R.id.my_view is used both as optional and required.*/R.id.my_view/**/) // THIS
                @Nullable View view;
            
                @OnClick(R.id.my_view)
                void click(View view) {}
            }
            """
        ).indented()
    )
    .issues(ButterKnifeConflictingOptionalityDetector.ISSUE_OPTIONALITY)
    .run()
    .expectInlinedMessages() // THIS
```

The API is defined in [lint-tests artifact](https://android.googlesource.com/platform/tools/base/+/studio-master-dev/lint/libs/lint-tests/src/main/java/com/android/tools/lint/checks/infrastructure/TestLintResult.kt#177)
See also https://proandroiddev.com/testing-your-first-android-lint-rule-bb78c1f2d2bd which shows another way, but I don't like because it depends on too much whitespace.

**Comment by TWiStErRob** [comment]
As an alternative/considerable rule, I would like to call out the approach we used at 
https://github.com/detekt/detekt/blob/b616a2f2420743e90cb49939b6284e5a8807a0b9/detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/BracesOnIfStatementsSpec.kt#L1979-L2010

Focus on the `"else"(3)`-style parts of the code, the builder mechanism definitely needs improving as you suggested as well. Explicitly listing the piece of code that's expected to be highlighted also solves the original problem you listed. Additionally it supports multiple occurrences, but not multiple different rules in the same `lint` result. It also keeps the code and the expected result separate, but close enough to be easily matched. Multi-line highlights could look strange, but those are discouraged anyway, because they would hide other issues in the highlighted code, so their numbers would be very limited.

**Comment by BraisGabin** [comment]
> I would add the requirement that at least 99% of the tests should be able to compile if their triple-quoted text is just copied out of the block into real code. Meaning I'm strongly against using emojis.

That would be a good feature, but I'm not so sure if it should be a requirement. I nearly never copy&paste the code that is inside the  `"""`. But I can see that it could happend some times and it would be helpful.

I like the android lint idea. They can inline the message so everything is inside the snippet. In my proposal that could be added later. A con is that, maybe, we don't want to define the message for all the snippets (right now nearlly no one test that).

What I like about emojis is that they are really easy to spot in long codes. The comments are not that easy to spot. Probably on `IDEA` is easier because it colors them. But on Github all is green text.

I'm not agains the Android lint idea, I want to have some different opinions so we can take a better decision.

---

The solution used in `BraceOnIfStatement` is nice but I think that the other 2 are better.

**Comment by TWiStErRob** [comment]
Just a quick note to record: if we overhaul the lint test API, and use a builder pattern, then the API could be designed to be extensible to allow all of these options if a third party wants to. And I think this might not be too much extra code.

**Comment by atulgpt** [comment]
We can also use `\*🐞*\` to represent the bug and at the same time, this will be compilable.

P.S. I do write the real code and then only put the code into the text(and sometimes vice-versa)

**Comment by cortinico** [comment]
Really like the idea!

> We can also use `/*🐞*/` to represent the bug and at the same time, this will be compilable.

+1 to this. If we start adding emojis inside the Kotlin code, we'll lose also the IntelliJ code injection feature, and we'll start seeing warnings inside the snippets that the Kotlin code is invalid.

On top of this: I would suggest to take a look at the Kotlin Compiler tests. They also follow a similar approach but use XML-like tags instead.

**Comment by BraisGabin** [comment]
> we'll lose also the IntelliJ code injection feature

I didn't think about that. That's a really good point 👍

**Comment by cortinico** [comment]
Is this something we want to see in 2.0 @BraisGabin ?

**Comment by BraisGabin** [comment]
It would be great. But lately I don't have that much time and I have other things with higher priority on my detekt list.

Also this should not be a breaking change so 2.0 shouldn't wait for this.

**Comment by cortinico** [comment]
> It would be great. But lately I don't have that much time and I have other things with higher priority on my detekt list.
> 
> Also this should not be a breaking change so 2.0 shouldn't wait for this.

Yup, I'll keep it in the backlog for now. Could be a nice clean-up for an hacktoberfest kind of issue though, if someone is up for preparing the infra to make this possible
